plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -146,21 +183,40 @@ private function process_img( OD_HTML_Tag_Processor $processor, OD_Tag_Visitor_C
 			$processor->remove_attribute( 'fetchpriority' );
 		}
 
-		// Ensure that sizes=auto is set properly.
-		$sizes = $processor->get_attribute( 'sizes' );
-		if ( is_string( $sizes ) ) {
+		// Ensure that sizes is set properly when it is a responsive image (it has a srcset attribute).
+		if ( is_string( $processor->get_attribute( 'srcset' ) ) ) {
+			$sizes = $processor->get_attribute( 'sizes' );
+			if ( ! is_string( $sizes ) ) {
+				$sizes = '';
+			}
+
 			$is_lazy  = 'lazy' === $this->get_attribute_value( $processor, 'loading' );
 			$has_auto = $this->sizes_attribute_includes_valid_auto( $sizes );
 
 			if ( $is_lazy && ! $has_auto ) {
-				$processor->set_attribute( 'sizes', "auto, $sizes" );
+				$new_sizes = 'auto';
+				if ( '' !== trim( $sizes, " \t\f\r\n" ) ) {
+					$new_sizes .= ', ';
+				}
+				$sizes = $new_sizes . $sizes;
 			} elseif ( ! $is_lazy && $has_auto ) {
 				// Remove auto from the beginning of the list.
-				$processor->set_attribute(
-					'sizes',
-					(string) preg_replace( '/^[ \t\f\r\n]*auto[ \t\f\r\n]*(,[ \t\f\r\n]*)?/i', '', $sizes )
-				);
+				$sizes = (string) preg_replace( '/^[ \t\f\r\n]*auto[ \t\f\r\n]*(,[ \t\f\r\n]*)?/i', '', $sizes );
 			}
+
+			// Compute more accurate sizes when it isn't lazy-loaded and sizes=auto isn't taking care of it.
+			if ( ! $is_lazy ) {
+				$computed_sizes = $this->compute_sizes( $context );
+				if ( count( $computed_sizes ) > 0 ) {
+					$new_sizes = join( ', ', $computed_sizes );
+					if ( '' !== $sizes ) {
```

Comment:
Should this be preserving the original sizes at the end? It can result in some conditions that are never matched. For example, if the original `sizes` attribute contains `(max-width: 600px) 480px, 800px` and then we prefix this value with:

```
(max-width: 480px) 432px, 
(min-width: 480px and max-width: 600px) 540px, 
(min-width: 600px and max-width: 782px) 704px, 
(min-width: 782px) 900px
```

Then the original conditions will never apply because `(min-width: 782px) 900px` is matched first. The only scenario where they could apply is if there aren't URL Metrics gathered for some of the viewport widths, and in this case leaving the original in place would serve as a fallback. Otherwise, we could remove it if all viewport groups are populated:

```suggestion
					if ( '' !== $sizes && $context->url_metric_group_collection->is_every_group_populated() ) {
```

But it doesn't seem to hurt to keep the original at the end.

Reply:
Wouldn't the media queries generated by Image Prioritizer always cover all viewport ranges? I don't see a reason to include the original sizes if they do. Of course if only some viewport groups are populated, it would be better to have the original sizes included.

That said, I think it would be safer to only alter the `sizes` attribute anyway when there's sufficient samples for all viewport groups. With the media queries it includes and the original sizes media queries, I'm wary of conflicts. Since we don't want to make anything worse but only better, I would be in favor to start with only computing `sizes` if all viewport groups are populated.

Reply:
They would only include viewport ranges for collected URL Metrics. So if nobody gets phablet or tablet traffic, then those viewport ranges would be missing. Nevertheless, the media features computed here include the minimum _and_ the maximum. So let's say that `sizes` attribute is originally `(max-width: 600px) 480px, 800px` from WordPress and then we prepend `(width <= 480px) 400px, (width > 872px) 800px` since we only have URL metrics for mobile and desktop. When these are combined with the original sizes, we get:

```
(width <= 480px) 400px, (width > 782px) 800px, (max-width: 600px) 480px, 800px
```

This means that, since `sizes` are processed left to right:

* Mobile viewports will get an image sized at 400px.
* Desktop viewports will get an image sized at 800px.
* Tablet and phablet visitors will fall back to the original sizes: `(max-width: 600px) 480px, 800px`, which means it will no worse than WordPress currently does by default.

Since processing happens left-to-right, and the first condition matched is used, then there won't be a conflict.

Reply:
Thanks for clarifying, that makes sense. May be worth adding an inline comment to mention that because of every viewport group providing minimum and maximum sizes, any viewport groups that are missing would simply still behave like before.

I like the idea of leaving out the original `sizes` value though if all viewport groups are populated, because then the original `sizes` would have no value at all and just look confusing in the frontend. If that's straightforward, let's add it.

Reply:
> Thanks for clarifying, that makes sense. May be worth adding an inline comment to mention that because of every viewport group providing minimum and maximum sizes, any viewport groups that are missing would simply still behave like before.

Done in d2b9a5c2c9503ee0e8965e0a506ac685460525d4

> I like the idea of leaving out the original `sizes` value though if all viewport groups are populated, because then the original `sizes` would have no value at all and just look confusing in the frontend. If that's straightforward, let's add it.

Done in 93961fa15faae40f1aeda3676ee2f501f5f2dd11. (Note my original suggestion above was incorrect in that I forgot to include the negation for the second condition, which I've now fixed in my suggestion.)

Note how now the `common-lcp-image-and-lazy-loaded-image-outside-viewport-with-fully-populated-sample-data` test case will omit the original `sizes` since all viewports are populated:

https://github.com/WordPress/performance/blob/d2b9a5c2c9503ee0e8965e0a506ac685460525d4/plugins/image-prioritizer/tests/test-cases/common-lcp-image-and-lazy-loaded-image-outside-viewport-with-fully-populated-sample-data/expected.html#L10

Compare this with the `only-mobile-and-desktop-groups-are-populated` test case which retains the original `sizes` as a fallback since the phablet and tablet URL Metrics are absent:

https://github.com/WordPress/performance/blob/d2b9a5c2c9503ee0e8965e0a506ac685460525d4/plugins/image-prioritizer/tests/test-cases/only-mobile-and-desktop-groups-are-populated/expected.html#L26